README.md

```diff
@@ -659,10 +659,290 @@ To begin coding the basic execution functionality of our VM, we only need these
+
+The method deduces the value of `Op0` if possible (based on `dst` and `Op1`).
+If Instruction opcode is a call `Op0` is deduced from the program counter and added the instruction size.
+If it is and `AssertEq` then a second switch case is used to check what `ResLogic` is, if it is `ResAdd` `Op0` is deduced from the substraction of `Op1` to `Dst` or if is is `Resmul` the `Op0` is deduced from the division of `Dst` and `Op1` (both felt values). otherwise op0 is nil.  
+
+```go
+func (vm *VirtualMachine) DeduceOp0(instruction *Instruction, dst *memory.MaybeRelocatable, op1 *memory.MaybeRelocatable) (deduced_op0 *memory.MaybeRelocatable, deduced_res *memory.MaybeRelocatable, error error) {
+	switch instruction.Opcode {
+	case Call:
+		deduced_op0 := vm.RunContext.Pc
+		deduced_op0.Offset += instruction.Size()
+		return memory.NewMaybeRelocatableRelocatable(deduced_op0), nil, nil
+	case AssertEq:
+		switch instruction.ResLogic {
+		case ResAdd:
+			if dst != nil && op1 != nil {
+				deduced_op0, err := dst.Sub(*op1)
+				if err != nil {
+					return nil, nil, err
+				}
+				return &deduced_op0, dst, nil
+			}
+		case ResMul:
+			if dst != nil && op1 != nil {
+				dst_felt, dst_is_felt := dst.GetFelt()
+				op1_felt, op1_is_felt := op1.GetFelt()
+				if dst_is_felt && op1_is_felt && !op1_felt.IsZero() {
+					return memory.NewMaybeRelocatableFelt(dst_felt.Div(op1_felt)), dst, nil
+
+				}
+			}
+		}
+	}
+	return nil, nil, nil
+}
+```
+
+##### DeduceOp1
+
+The method deduces the value of `Op1` if possible (based on `dst` and `Op0`) it also deduces `res` if possible.
+If the instruction opcode is `AssertEq` a switch case is used to check what the `ResLogic` is, if it is a `ResOp1` then the value of op1 is equal to the dst operand. If `ResLogic` is `ResAdd` op1 is deduced from the substraction of `dst` with `op0`. Finally if it is `ResMul` `op1` is deduced from the division of `dst` with `op0`, in all the cases `res` is equal to `dst`. if none of the former cases apply then nil is returned.
+
+```go
+func (vm *VirtualMachine) DeduceOp1(instruction Instruction, dst *memory.MaybeRelocatable, op0 *memory.MaybeRelocatable) (*memory.MaybeRelocatable, *memory.MaybeRelocatable, error) {
+	if instruction.Opcode == AssertEq {
+		switch instruction.ResLogic {
+		case ResOp1:
+			return dst, dst, nil
+		case ResAdd:
+			if op0 != nil && dst != nil {
+				dst_rel, err := dst.Sub(*op0)
+				if err != nil {
+					return nil, nil, err
+				}
+				return &dst_rel, dst, nil
+			}
+		case ResMul:
+			dst_felt, dst_is_felt := dst.GetFelt()
+			op0_felt, op0_is_felt := op0.GetFelt()
+			if dst_is_felt && op0_is_felt && !op0_felt.IsZero() {
+				res := memory.NewMaybeRelocatableFelt(dst_felt.Div(op0_felt))
+				return res, dst, nil
+			}
+		}
+	}
+	return nil, nil, nil
+}
+
 ```
+##### ComputeRes
+
+If the Res value has not been deduced in the previous steps then it is computed based on the `Op0` and `Op1` values and the res logic stored in the instruction.
```

Comment:
What do you mean with `res logic`?

Reply:
is how it is called in the Instruction struct. i will change it.